File: prog_lp/py/Conjunto.py
class Conjunto:
    def __init__(A):
        A.lista = []

    def ehVazio(A):
        return not A.lista
    
    def uniaoConj(A, B):
        C = Conjunto()
        for i in A.lista:
            C.addConj(i)
        for j in B.lista:
            if j not in C.lista:
                C.addConj(j)
        C.lista.sort()
        return C
    
    def intersConj(A, B):
        C = Conjunto()
        for i in A.lista:
            if i in B.lista:
                C.addConj(i)
        C.lista.sort()
        return C

    def diffConj(A, B):
        C = Conjunto()
        for i in A.lista:
            if i not in B.lista: 
                C.addConj(i)
        C.lista.sort()
        return C
    
    def diffSimConj(A, B):
        uniao = A.uniaoConj(B)
        intersecao = A.intersConj(B)
        C = uniao.diffConj(intersecao)
        C.lista.sort()
        return C

    def addConj(A, x):
        if x not in A.lista:
            A.lista.append(x)

    def remConj(A, x):
        if x in A.lista:
            A.lista.remove(x)

    def tamConj(A):
        return len(A.lista)

    def ehElemConj(A, x):
        return x in A.lista

File: linguagens_programacao/prolog/conjunto.pl
% Verifica se o conjunto é vazio
ehVazio([], true).
ehVazio(_, false).

% União de conjuntos
uniaoConj(A, B, Uniao) :-
    append(A, B, Uniao).

% Interseção de conjuntos
intersConj([], _, []).
intersConj([X|RestoA], B, Intersecao) :-
    (   member(X, B)
    ->  Intersecao = [X|RestoIntersecao]
    ;   Intersecao = RestoIntersecao
    ),
    intersConj(RestoA, B, RestoIntersecao).

% Diferença de conjuntos
diffConj([], _, []).
diffConj([X|RestoA], B, Diferenca) :-
    (   member(X, B)
    ->  Diferenca = RestoDiferenca
    ;   Diferenca = [X|RestoDiferenca]
    ),
    diffConj(RestoA, B, RestoDiferenca).

% Diferença simétrica de conjuntos
diffSimConj(A, B, DiferencaSimetrica) :-
    diffConj(A, B, DiferencaAB),
    diffConj(B, A, DiferencaBA),
    append(DiferencaAB, DiferencaBA, DiferencaSimetrica).

% Adiciona elemento ao conjunto
addConj(Conj, X, NovoConj) :-
    (   member(X, Conj)
    ->  NovoConj = Conj
    ;   NovoConj = [X|Conj]
    ).

% Remove elemento do conjunto
remConj([], _, []).
remConj([X|Resto], Y, NovoConj) :-
    (   X = Y
    ->  NovoConj = Resto
    ;   NovoConj = [X|RestoNovoConj]
    ),
    remConj(Resto, Y, RestoNovoConj).

% Retorna o tamanho do conjunto
tamConj(Conj, Tamanho) :-
    length(Conj, Tamanho).

% Verifica se um elemento pertence ao conjunto
ehElemConj(Conj, X) :-
    member(X, Conj).
